retry.go

```diff
@@ -0,0 +1,70 @@
+package rueidis
+
+import (
+	"context"
+	"math"
+	"time"
+
+	"github.com/redis/rueidis/internal/util"
+)
+
+const (
+	defaultMaxRetryDelay = 1 * time.Second
+	maxAttemptsShift     = 63 // Avoid excessive shifts
+)
+
+// RetryDelay returns the delay that should be used before retrying the
+// attempt. Will return negative delay if the delay could not be determined or do not retry.
+type RetryDelay func(attempts int, err error) time.Duration
+
+// defaultRetryDelay delays the next retry exponentially without considering the error.
+// max delay is 1 second.
+func defaultRetryDelay(attempts int, _ error) time.Duration {
+	base := util.FastRandFloat64()
+	backoff := uint64(1 << uint64(math.Min(float64(attempts), maxAttemptsShift)))
+	jitter := base * float64(backoff) * float64(time.Millisecond)
+	return time.Duration(math.Min(jitter, float64(defaultMaxRetryDelay)))
+}
+
+// retryableCheckFunc is a function that checks if the command should be retried.
+// Returns true if the command should be retried.
+type retryableCheckFunc func() bool
+
+type retryHandler interface {
+	// WaitUntilNextRetry waits until the next retry should be attempted.
+	// Returns false without error immediately if the command should not be retried.
+	// Returns false and an error if waiting for the next retry was interrupted.
+	// Returns true after the delay if the command should be retried.
+	WaitUntilNextRetry(ctx context.Context, isRetryable retryableCheckFunc, attempts int, err error) (bool, error)
+}
+
+type retryer struct {
+	RetryDelay RetryDelay
+}
+
+func newRetryer(retryDelay RetryDelay) *retryer {
+	return &retryer{RetryDelay: retryDelay}
+}
+
+func (r *retryer) WaitUntilNextRetry(
+	ctx context.Context, isRetryable retryableCheckFunc, attempts int, err error,
+) (bool, error) {
+	if !isRetryable() {
+		return false, nil
+	}
+
+	delay := r.RetryDelay(attempts, err)
+	if delay < 0 {
+		return false, nil
+	}
+
+	timer := time.NewTimer(delay)
+	defer timer.Stop()
+
+	select {
+	case <-ctx.Done():
```

Comment:
We should not wait for the context deadline but check whether the deadline will be exceeded or not:

```go
	dl, ok := ctx.Deadline()
	if ok && time.Until(dl) > delay {
		time.Sleep(delay)
	}
```

Reply:
Is there no need to respect context cancel ?

Reply:
It is better to respect the cancellation, but the returned `errAbortWaiting` complicates the code outside.

How about we do this, removing all the return values:

```go
	ctx := context.Background()
	if dl, ok := ctx.Deadline(); !ok || time.Until(dl) > delay {
		if ch := ctx.Done(); ch != nil {
			tm := time.NewTimer(delay)
			defer tm.Stop()
			select {
			case <-ch:
			case <-tm.C:
			}
		} else {
			time.Sleep(delay)
		}
	}
```

Reply:
We still let it retry if the context is ended. The underlying pipe will check the ctx.Err().

Reply:
outside code of `WaitUntilNextRetry` can be complicated. 

I feel that this is bit more code style. when not returning error, caller should always check context is done or ,as you can said, `pipe` always will check and all contributors should know that. It is unclear to me, so that complicated but explicit code is more sense to me. I change it, but can you give me more detailed reason?

Reply:
[ea85114](https://github.com/redis/rueidis/pull/643/commits/ea8511464d7b71612c061b9366380764138303cc)

Reply:
> pipe always will check and all contributors should know that. It is unclear to me

It is indeed implicit. But the reason for not returning an error is simple: adding a new error path to replace a `goto retry` statement seems doing too much.